src/PluginsManager/Microsoft.Performance.Toolkit.Plugins.Runtime/Manager/PluginsManager.cs

```diff
@@ -0,0 +1,624 @@
+            IEnumerable<IPluginFetcher> fetchers,
+            IPluginRegistry pluginRegistry,
+            string installationDir,
+            ILogger logger)
+        {
+            Guard.NotNull(discovererProviders, nameof(discovererProviders));
+            Guard.NotNull(fetchers, nameof(fetchers));
+            Guard.NotNull(pluginRegistry, nameof(pluginRegistry));
+            Guard.NotNullOrWhiteSpace(installationDir, nameof(installationDir));
+            Guard.NotNull(logger, nameof(logger));
+
+            this.discovererRepository = new PluginsManagerResourceRepository<IPluginDiscovererProvider>(discovererProviders);
+            this.fetcherRepository = new PluginsManagerResourceRepository<IPluginFetcher>(fetchers);
+            this.discovererSourcesManager = new DiscovererSourcesManager(this.discovererRepository);
+            this.discovererSourcesManager.PluginSourceErrorOccured += (s, e) => PluginSourceErrorOccured?.Invoke(s, e);
+
+            this.pluginRegistry = pluginRegistry;
+            this.pluginInstaller = new PluginInstaller(pluginRegistry);
+            this.installationDir = installationDir;
+            this.logger = logger;
+        }
+
+        /// <inheritdoc />
+        public event EventHandler<PluginSourceErrorEventArgs> PluginSourceErrorOccured;
+
+        /// <inheritdoc />
+        public IEnumerable<PluginSource> PluginSources
+        {
+            get
+            {
+                return this.discovererSourcesManager.PluginSources;
+            }
+        }
+
+        /// <inheritdoc />
+        public void ClearPluginSources()
+        {
+            this.discovererSourcesManager.ClearPluginSources();
+        }
+
+        /// <inheritdoc />
+        public void AddPluginSources(IEnumerable<PluginSource> sources)
+        {
+            this.discovererSourcesManager.AddPluginSources(sources);
+        }
+
+        // TODO: Re-enable when we start to support loading additional resources.
+        ///// <inheritdoc />
+        //public bool LoadAdditionalPluginResources(string directory)
+        //{
+        //    return this.resourceLoader.TryLoad(directory);
+        //}
+
+        /// <inheritdoc />
+        public async Task<IReadOnlyCollection<AvailablePlugin>> GetAvailablePluginsLatestAsync(
+            CancellationToken cancellationToken)
+        {
+            PluginSource[] pluginSources = this.PluginSources.ToArray();
+            Task<IReadOnlyCollection<AvailablePlugin>>[] tasks = this.PluginSources
+                .Select(s => GetAvailablePluginsLatestFromSourceAsync(s, cancellationToken))
+                .ToArray();
+
+            var task = Task.WhenAll(tasks);
+            try
+            {
+                await task.ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                this.logger.Info($"The request to get lastest available plugins is cancelled.");
+                throw;
+            }
+            catch
```

Comment:
Should we catch this exception too?

Reply:
Since this is running in a Task.WhenAll, this catch here won't be able to handle exceptions from each individual task. They are instead handled in the for loop below. Same for other places that use Task.WhenAll.